control-plane/typeSpecConverter.cpp

```diff
@@ -111,6 +115,48 @@ bool TypeSpecConverter::preorder(const IR::Type_Name* type) {
     return false;
 }
 
+// This function is invoked if psa.p4 has a Newtype. The function
+// is not invoked for user-defined NewType.
+bool TypeSpecConverter::preorder(const IR::Type_Newtype* type) {
+    if (p4RtTypeInfo) {
+        bool orig_type = true;
+        const IR::StringLiteral* uri;
+        const IR::Constant* sdnB;
+        auto ann = type->getAnnotation("p4runtime_translation");
+        if (ann != nullptr) {
+            orig_type = false;
+            uri = ann->expr[0]->to<IR::StringLiteral>();
+            BUG_CHECK(uri, "P4runtime annotation does not have uri: %1%",
+                      type->getP4Type());
+            sdnB = ann->expr[1]->to<IR::Constant>();
+            BUG_CHECK(sdnB, "P4runtime annotation does not have sdn: %1%",
+                      type->getP4Type());
+        }
+
+        auto name = std::string(type->controlPlaneName());
+        auto types = p4RtTypeInfo->mutable_new_types();
+        if (types->find(name) == types->end()) {
+            auto newTypeSpec = new p4configv1::P4NewTypeSpec();
+            auto newType = type->type;
+            auto n = newType->to<IR::Type_Name>();
+            visit(n);
```

Comment:
I don't understand this, but it looks wrong. We very seldom need to recursively call a visitor.
Why is n not nullptr?
I expect this will fail if the same Type_Name appears multiple times in the program.

Reply:
> I don't understand this, but it looks wrong. We very seldom need to recursively call a visitor.
> Why is n not nullptr?

First, `TypeSpecConverter::preorder(const IR::Type_Name* type)` is called.  This function has processed the `type->path` in refMap to get the `decl`.  The decl type is NewType who type is a Type_Name.  The same Type_Name is `n` in this code and is not a nullptr.

There is no recursion. A Newtype is found and its Type_Name is sent to `visit` to emplace in the map.  Then typeSpec is obtained from map.at(n).  The same two-step process is used by many preorder methods in this file. 

> I expect this will fail if the same Type_Name appears multiple times in the program.

This code is processing psa.p4 which has PortId_t uses multiple times in the program with no problems found.

Reply:
First, the Type_Name is terminal. It has no Type_Bits to traverse to.  As I said above, the visit(n) emplaces the Type_Name in the map (defined in typeSpecConverter.h).

Second, I deliberately set n to nullptr before visit(n) and visit does not have a problem.

Reply:
Visit should handle nullptrs (and anything else) just fine.  You need to be a bit careful calling visit in a Modifier or Transform like this, but for an Inspector, the only thing you need to worry about is triggering a recursive traversal (which will trigger an "IR loop detected" BUG if it happens.)